src/ApplicationInsights.Kubernetes/K8sInfoBootstrap.cs

```diff
@@ -0,0 +1,87 @@
+using Microsoft.ApplicationInsights.Kubernetes.Debugging;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.ApplicationInsights.Kubernetes;
+
+internal class K8sInfoBootstrap : IK8sInfoBootstrap
+{
+    private readonly ApplicationInsightsKubernetesDiagnosticSource _logger = ApplicationInsightsKubernetesDiagnosticSource.Instance;
+
+    private readonly AppInsightsForKubernetesOptions _options;
+
+    private readonly object _locker = new object();
+    private readonly IServiceScopeFactory _serviceScopeFactory;
+    private bool _hasExecuted = false;
+
+    public K8sInfoBootstrap(
+        IServiceScopeFactory serviceScopeFactory,
+        IOptions<AppInsightsForKubernetesOptions> options)
+    {
+        _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+        _serviceScopeFactory = serviceScopeFactory ?? throw new ArgumentNullException(nameof(serviceScopeFactory));
+    }
+
+    public async Task ExecuteAsync(CancellationToken cancellationToken)
+    {
+        // Simple mechanism to making sure it is safe to call into this method multiple times.
+        lock (_locker)
+        {
+            if (_hasExecuted)
+            {
+                return;
+            }
+            _hasExecuted = true;
+        }
+
+        TimeSpan interval = _options.ClusterInfoRefreshInterval;
+
+        while (!cancellationToken.IsCancellationRequested)
+        {
+            try
+            {
+                _logger.LogDebug($"Starting update K8sEnvironment");
+
+                using (IServiceScope scope = _serviceScopeFactory.CreateScope())
+                {
+                    IK8sEnvironmentFetcher fetcher = scope.ServiceProvider.GetRequiredService<IK8sEnvironmentFetcher>();
+                    await fetcher.UpdateK8sEnvironmentAsync(cancellationToken);
+                }
+
+                _logger.LogDebug($"Finished update K8sEnvironment, next iteration will happen at {DateTime.UtcNow.Add(interval)} (UTC) by interval settings of {interval}");
+
+                // TODO: Consider using PeriodicTimer than delay below when move to .NET 6 +
+                await Task.Delay(interval, cancellationToken: cancellationToken);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception ex)
+            {
+                // In case of any unexpected exception, we shall log it than let it throw.
+                _logger.LogError("Unexpected error happened. Telemetry enhancement with Kubernetes info won't happen. Message: {0}", ex.Message);
+                _logger.LogTrace("Unexpected error happened. Telemetry enhancement with Kubernetes info won't happen. Details: {0}", ex.ToString());
+            }
+#pragma warning restore CA1031 // Do not catch general exception types
+        }
+    }
+
+
+    // Allows non-hosted service to bootstrap cluster property fetch.
+    public void Run()
```

Comment:
Consider taking a `CancellationToken` parameter with default value equal to `default` or `CancellationToken.None`

Reply:
Done.